### ISSUE TITLE
Cannot find module './drivers/node-mongodb-native/connection' using Parcel

### ISSUE THREAD

**Issue by julesmons** [Automattic/mongoose#9603]
**Do you want to request a *feature* or report a *bug*?**
*bug*

**What is the current behavior?**
I'm using parcel v2 to bundle an entire node api into one JS file. 
It's a relatively small api, and i need to bundle all of it because of the hosting env it will ultimately be placed on.

The API uses mongoose and when it's bundled & 'compiled' by parcel it doesn't raise any errors. However, when running the application it doesn't work because certain parts of mongoose are missing, that's where the error "Cannot find module './drivers/node-mongodb-native/connection'" comes from. _The runtime environment can't find this module because it's never bundled._

The cause of this problem has been found in other issues from other libraries. See:
https://github.com/parcel-bundler/parcel/issues/4031
https://github.com/evanw/esbuild/issues/480

The esbuild issue is from a different library, but it boils down to the same problem. 
Some parts of mongoose are required dynamically using a variable. See: https://github.com/Automattic/mongoose/blob/1c2978e31f4c46be135e6ee69e668027ceb4bf5e/lib/index.js#L729-L735

The bundlers don't pick this up and this results in those parts not being bundled, and therefore not being available at runtime.

**If the current behavior is a bug, please provide the steps to reproduce.**

To reproduce this issue, i've included a sample project below. Just use "npm run start" to see the error.

<details>
    <summary>The source files</summary>

    #### package.json
    ```json
    {
        "name": "test",
        "version": "1.0.0",
        "description": "",
        "author": "Jules Mons",
        "license": "ISC",
        "main": "./dist/js/main.js",
        "source": "./src/js/main.js",
        "targets": {
            "main": {
                "distDir": "./dist",
                "context": "node",
                "engines": {
                    "node": "12"
                },
                "includeNodeModules": true
            }
        },
        "scripts": {
            "start": "npm-run-all build serve",
            "serve": "node .",
            "build": "cross-env NODE_ENV=production parcel build ./src/js/main.js --target main"
        },
        "dependencies": {
            "mongoose": "5.10.15"
        },
        "devDependencies": {
            "npm-run-all": "4.1.5",
            "parcel": "^2.0.0-nightly.460"
        }
    }
    ```

    #### src/js/main.js
    ```javascript
    import mongoose from "mongoose";

    mongoose.connect(

        "mongodb://localhost:27017", 
        { useNewUrlParser: true }, 
        (err) => {

            if (!err) {

                console.log("Connection established.");
            }
        }
    );
    ```
</details>

**What is the expected behavior?**
I would expect to see all modules that are needed to be bundled. 

A way to achieve this would be to not use a dynamic require. I saw the global.MONGOOSE_DRIVER_PATH is being deprecated anyway. Maybe this issue could be addressed in the next release?

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Software | Version(s)
-- | --
Parcel | 2.0.0-nightly.460
Mongoose | 5.10.15
Node | 15.3.0, api is compiled to 12
npm | 7.0.8
Operating System | Windows 10

**Comment by vkarpov15** [comment]
We definitely want to remove MONGOOSE_DRIVER_PATH in the next major release. However, removing it before then is very risky and may break other people's code because MONGOOSE_DRIVER_PATH has been supported for about 8 years at this point. We use it internally to make sure we avoid the node driver so we can build Mongoose's browser library.

Until then, you'll have to find a workaround with parcel to pull in the missing files. I don't know enough about parcel yet to suggest a workaround, but I'll look into it later this week.

**Comment by julesmons** [comment]
Thank you! Parcel isn't able to manually require the missing files. Nonetheless, thanks for picking this up!

**Comment by stolinski** [comment]
Hitting this with https://esbuild.github.io

**Comment by hugdru** [comment]
I am using esbuild through [aws-lambda-nodejs](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-nodejs-readme.html). Noticed some warnings due to dynamic imports which resulted in runtime errors. For logform I followed https://github.com/evanw/esbuild/issues/480#issuecomment-715671679 but for mongoose I used [patch-package](https://www.npmjs.com/package/patch-package) to modify:

<details>
  <summary>/lib/index.js</summary>

https://github.com/Automattic/mongoose/blob/1c2978e31f4c46be135e6ee69e668027ceb4bf5e/lib/index.js#L7-L16
https://github.com/Automattic/mongoose/blob/1c2978e31f4c46be135e6ee69e668027ceb4bf5e/lib/index.js#L725-L741

</details>

It is not ideal but works. I also extended [NodejsFunction](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda-nodejs.NodejsFunction.html) to avoid having to apply these workarounds for each lambda that needs them using spawnsync, fs, and hooks.

**Comment by CryogenicPlanet** [comment]
> I am using esbuild through [aws-lambda-nodejs](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-nodejs-readme.html). Noticed some warnings due to dynamic requires which resulted in runtime errors. For logform I followed [evanw/esbuild#480 (comment)](https://github.com/evanw/esbuild/issues/480#issuecomment-715671679) but for mongoose I used [patch-package](https://www.npmjs.com/package/patch-package) to modify:
> /lib/index.js
> 
> It is not ideal but works. I also extended [NodejsFunction](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda-nodejs.NodejsFunction.html) to avoid having to apply these workarounds for each lambda that needs them. Used hooks, fs, and spawnsync in the implementation.

@hugdru Could you share your patched code? I am having the same problem with `esbuild` and just blocked on it

**Comment by hugdru** [comment]
@CryogenicPlanet Yeah sure. I followed the steps in [patch-package](https://www.npmjs.com/package/patch-package) which resulted in _patches/mongoose+5.10.19.patch_ in the root of the project.

https://gist.github.com/hugdru/982a31d929a030e29f852c4bfafd9b16

**Comment by raphaelschwinger** [comment]
@CryogenicPlanet for your suggested changes! We could use those in our project. Since we are also effected by https://github.com/typegoose/typegoose/issues/480 and therefore need mongoose version `5.10.18` I forked this repo and import a [branch](https://github.com/raphaelschwinger/mongoose/tree/fix-import) with your patch applied and import it with [yarn](https://yarnpkg.com/cli/add). 
```
    "mongoose": "raphaelschwinger/mongoose#fix-import"
```

**Comment by vkarpov15** [comment]
I took a closer look at this and we may be able to work around this in Mongoose 5.x without breaking the API completely. We'll review a bit more but otherwise we should be able to release this fix with Mongoose 5.12.8 :+1:

**Comment by vkarpov15** [comment]
Fix will be in 5.12.8 :+1:
